facebook/ThreatExchange
# [py-tx][backwards incompatible] Split Update Format and Index Format

Summary
---------
The flow of information through the system is:

API.fetch() => Delta => Store => Index => Metadata

NCMEC's API is like the tag-based ThreatExchange API, and it's not stored as hash => record, it's actually stored record => [hash, hash, hash] which is incompatible with many assumptions made along the way.

After lots of mulling, I think the solution is to:
1. Separate considerations about storing the record that the fetch() API produces
2. Converting the format from the fetch() update store to the signal => metadata

As a part of that, I need to mess with a lot of the interfaces above. There are more changes that I want to make, and I tried to checkpoint with this diff.

The final version I want to end up with is:
1. Storing what the update format is from fetch should move from Delta => SignalExchangeAPI
2. Delta should become a trivial object, with no need to override it

As a side effect, the storage format for the CLI changed from JSON to pickle. It's possible I can actually revert this in a few more PRs, but right now Delta is the only object that knows the correct format.


Test Plan
---------

Added a test

Also went through:
1. Fetch from threatexchange
3. Match some content
4. Run the dataset command to make sure it isn't broken (yet).
